src/util/memstream.rs

```diff
@@ -0,0 +1,159 @@
+// Copyright 2018 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under the MIT license <LICENSE-MIT
+// https://opensource.org/licenses/MIT> or the Modified BSD license <LICENSE-BSD
+// https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be copied,
+// modified, or distributed except according to those terms. Please review the Licences for the
+// specific language governing permissions and limitations relating to use of the SAFE Network
+// Software.
+
+//! Tokio compatible in-memory stream implementation. Helpful for testing.
+
+// TODO(povilas): move to future_utils crate - this module is generic and highly reusable.
+
+use priv_prelude::*;
+use std::fmt;
+use std::io::{self, Read, Write};
+
+/// In-memory stream implementing `AsyncWrite + AsyncRead`. This stream is not connected with
+/// any other stream. Meaning what you write to the stream you can later read from it.
+#[derive(Default)]
+pub struct EchoStream {
+    buf: Vec<u8>,
+    shutdown: bool,
+}
+
+impl Read for EchoStream {
+    /// Reads as much data as possible to the given buffer and removes that data from internal
+    /// buffer.
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if self.buf.is_empty() {
+            if self.shutdown {
+                return Ok(0);
+            } else {
+                return Err(io::ErrorKind::WouldBlock.into());
+            }
+        }
+
+        let mut bytes_read = 0;
+        while bytes_read < buf.len() && !self.buf.is_empty() {
+            buf[bytes_read] = self.buf.remove(0);
+            bytes_read += 1;
+        }
+        Ok(bytes_read)
+    }
+}
+
+impl Write for EchoStream {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.buf.extend_from_slice(buf);
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl AsyncRead for EchoStream {}
+
+impl AsyncWrite for EchoStream {
+    /// Analogous to sending FIN to myself.
+    /// This will unblock `read()` calls which will get 0 as result value.
+    fn shutdown(&mut self) -> io::Result<Async<()>> {
+        self.shutdown = true;
+        Ok(Async::Ready(()))
+    }
+}
+
+impl fmt::Debug for EchoStream {
```

Comment:
is it necessary ? the only extra field you are avoiding to print happens to be a bool. Does it make debugging a lot easier to avoid that forcing us to hand code a debug impl ?

Reply:
Crust required `PaStream` variants to impl `Debug`, not sure why, will check,

Reply:
just to make it clear(er) just in case, i meant was manual impl is not needed - the default one should suffice. I'm not saying to get rid of `Debug` trait for `PaStream`.